### ISSUE TITLE
The "<100 images" low-data claim isn't reproduced by the training code — proposal for a data-scaling study

### ISSUE THREAD

**Issue by Kuo-TingKai** [ChiShengChen/Quantum_MNIST_Diffusion#5]
## Summary

The README makes what I think is the most interesting claim in the paper:

> **We trained the quantum diffusion model with fewer than 100 images, demonstrating the advantage of quantum layers in low-data regimes.**

I can't find the code path that produces this result, and the training code as committed appears to use the full class subset. I'd like to (a) ask where the <100-image experiment lives, and (b) propose turning this one-point claim into a scaling study, which I think is where the publishable contribution actually is.

## What the committed code trains on

`quantum_difussion_mnist_v7.py`:

```python
dataset = datasets.MNIST(root="./data", train=True, download=True, transform=transform)
dataset.data = dataset.data[dataset.targets == digit_label]
dataset.targets = dataset.targets[dataset.targets == digit_label]
loader = DataLoader(dataset, batch_size=64, shuffle=True)
```

That's the entire MNIST training split for one digit — roughly **5 400–6 700 images** depending on the class, not <100. There's no subsampling anywhere in `train_pipeline`, and no `n_train` / `--subset` argument. `full_unet/quantum_diffusion_mnist_v8.py` does the same.

If the <100-image runs were done in a separate script that didn't make it into the repo, adding it would help a lot — it's the headline claim.

## Why a scaling study is the stronger paper

A single "quantum wins at n<100" data point is exactly the kind of result reviewers discount, because it's one draw from a high-variance estimator. But the same experiment run as a curve is a genuine contribution, and it connects directly to existing theory: Caro et al., *"Generalization in quantum machine learning from few training data"* (Nature Communications 13, 4919, 2022) predicts generalization error scaling like $\sqrt{T/N}$ in the number of trainable gates $T$, which makes a concrete, falsifiable prediction about **where** the crossover should be. Very few QML generative papers actually test a scaling prediction.

Concretely:

* **x-axis:** $N \in \{10, 25, 50, 100, 250, 500, 1000, \text{full}\}$ training images per class.
* **arms:** the four-way ablation ladder in #3 (plain / classical-SE / frozen-random-SE / quantum), so a crossover can be attributed to the circuit rather than to the gating.
* **seeds:** ≥5 per (N, arm) cell, with the data subset resampled per seed. In the $N=10$ regime, *which* 10 images you draw will dominate everything else.
* **y-axis:** KID (unbiased at small $N$) plus Precision/Recall — see #4 for why the current FID numbers can't support this comparison.
* **report:** mean ± CI, and a paired test across seeds. The headline plot is "gap (quantum − classical-SE) vs N, with the crossover point marked".

The result is publishable **either way**. If the crossover exists and is robust across seeds, that's a real low-data claim with theory behind it. If it doesn't, "the apparent low-data quantum advantage in hybrid diffusion models disappears under a parameter-matched classical ablation" is a useful negative result — that genre gets accepted (cf. the classical-baseline literature around quantum kernels).

## Practical notes

* Fixing #2's per-sample Python loop (v7 runs one QNode call per batch element) is close to a prerequisite — this grid is ~8 × 4 × 5 = 160 runs per digit.
* With $N=10$, 30 epochs is ~5 gradient steps. The epoch budget needs to be defined in *steps*, not epochs, or the low-N arms are simply undertrained rather than data-limited.
* Related: `quantum_difussion_mnist_v7.py`'s main block is `for digit in range(1, 10)`, so digit **0** is never trained by that script, though the README shows digit-0 GIFs and the metrics CSVs include a digit-0 row. Worth double-checking which script produced those.

## Two smaller things that block reproduction of any of this

* **No seeding.** There's no `torch.manual_seed` / `np.random.seed` anywhere, so no run in the repo is reproducible and no reported difference has an error bar.
* **Model selection on a single-batch loss.** In the training loop, `loss_history.append(loss.item())` records only the *last* minibatch of the epoch, and `if loss.item() < best_loss` uses that same single-batch value to decide which checkpoint to keep. With batch size 64 and MSE over random timesteps, that value is dominated by which $t$'s happened to be sampled in that batch, so `best_model.pth` is close to a random epoch. An epoch-mean would make both the loss curve and the checkpoint meaningful.

**Comment by ChiShengChen** [comment]
Answering (a) first: **there is no <100-image code path.** You read the code correctly — `train_pipeline` trains on the whole class with no subsampling and no way to request any. I've retracted the claim from the README rather than leave it standing while the experiment is built.

Also confirmed your digit-0 observation: `quantum_difussion_mnist_v7.py` ended in `for digit in range(1, 10)`, so **digit 0 was never trained by that script**. The digit-0 rows in the committed CSVs are under `full_unet/`, and v8's main block is `range(10)` — so those came from v8, not v7. `--digits` now defaults to 0–9.

The machinery for (b) is in #9:

- `--n-train N`, with the subset **resampled per seed** — at N=10 which ten images you draw dominates everything else, exactly as you said.
- `--max-steps`, a budget in optimizer steps. You were right that this is load-bearing and not a detail: at N=10 an epoch is one gradient step.
- `run_scaling_study.py` enumerates the (N, arm, seed) grid.
- Per-run `manifest.json` + `loss.csv` so a cell is traceable.

**The thing you'll want to know before planning this:** the grid as specified costs about **508 hours serial, 503 of which is the quantum arm**.

```
$ python run_scaling_study.py --digits 3 --dry-run
160 cells: 1 digit(s) x 8 N x 4 arm(s) x 5 seed(s)
rough serial cost: 508.3 h (21.2 days) on one CPU core, dominated by the quantum arm
  quantum arm alone: 503.3 h of that
```

That's one digit. The quantum arm is ~350× slower *per step* than the classical arms because it backpropagates through a 2^16 statevector, and #7 made this worse (correctly — the old one wasn't computing a real gradient at all).

One lever that helps: **`lightning.qubit` + `adjoint` is 3.5× faster** than `default.qubit` + `backprop` (6.22 s vs 21.97 s per step at n_qubits=16, batch 64), and I verified the gradients agree to ~5e-07 on a scale of ~1, so it's not a trade-off. Exposed as `--qdevice` / `--diff-method`, with a test pinning the equivalence. That takes the quantum arm to ~144 h, which is a sweep you can actually run.

Beyond that the honest options are fewer N values, fewer qubits, or a GPU statevector backend (`lightning.gpu` isn't installed here). Happy to benchmark a narrower grid if you want to pick a shape.

Two smaller prerequisites from #9 that matter for the sweep specifically: `best_model.pth` was being selected on a single batch's loss (now the epoch-window mean), and the EMA model was built with `copy.deepcopy`, which raises for any non-default simulator.

**Comment by ChiShengChen** [comment]
Follow-up: the step budget is now on all three scripts, and the PathMNIST path is verified rather than assumed.

**Step budget ported to v8 and pathmnist.** Both were epoch-based, so their cells at different `--n-train` weren't compute-matched — exactly the confound you flagged. Both now take `--max-steps` with the same window-mean checkpointing as v7; `--epochs` still works when `--max-steps` is omitted. `run_scaling_study.py` emits a budget for all three now.

While porting, pathmnist turned out to have the **same checkpoint-selection bug v7 had** — `best_loss` was the *last batch's* loss of each epoch, not the epoch mean. It also sampled every epoch (1000 reverse steps each), which was a large fraction of its runtime; that's now `--sample-every`, default off.

**PathMNIST is now actually run, not just read.** I installed medmnist, so:

- `--arm se` and `--arm quantum` both train end to end and write `[5, 3, 28, 28]` samples in `[-1, 1]`, which is what `evaluate.py` expects.
- `--qdevice lightning.qubit --diff-method adjoint` works there too.
- The fast label filter was checked against the old one **on the real dataset**: identical indices (9 509 of 89 996 for label 1), 4.7 s → 0.4 ms. There's a test asserting the two agree, not just that they return the same count.

One correction to what I said earlier: I described the old label filter as "minutes of work". It's **4.7 s** per call — still ~11 000× more than needed, and it runs once per cell, but not minutes. PathMNIST images are already 28×28, so the Resize is cheap.

Test suite went from 66 passed / 13 skipped to **79 passed / 3 skipped** — the pathmnist tests were skipping for want of medmnist and now actually run.

The cost picture from before is unchanged and is still the thing that decides whether this study is runnable: ~508 h serial for one digit, ~503 h of it the quantum arm, reducible to ~144 h with `lightning.qubit` + `adjoint`. If you tell me which N values and how many seeds you actually want, I can price that specific shape.

**Comment by ChiShengChen** [comment]
Ran the sweep. The headline result is about **training budget, not data quantity** — and it changes how this study should be designed.

## 2000 steps produces noise regardless of data

I ran the full ladder at 2000 steps (roughly the repo's original 30 epochs × ~6000 images), 8 values of N, 5 seeds, 120 cells. KID was flat across the entire range:

| N | 10 | 100 | 1000 | 6131 |
|---|---|---|---|---|
| `plain` KID | 0.513 | 0.506 | 0.478 | 0.452 |

Against a real-vs-real floor of **KID 0.00021**. A 600× increase in data produced no measurable improvement, because at this budget the model isn't learning to generate digits at all — so data quantity is not the binding constraint.

## At a real budget, the picture inverts

| arm | steps | KID | FID |
|---|---|---|---|
| `plain` | 2 000 | 0.463 | — |
| `plain` | 30 000 | 0.453 | 313.8 |
| `plain` | 100 000 | 0.310 | 257.2 |
| `se` | 2 000 | 0.479 | 319.1 |
| `se` | 10 000 | 0.181 | 146.6 |
| `se` | **30 000** | **0.103 ± 0.006** (3 seeds) | **76.8** |

Two things follow:

1. **The bottleneck is a large efficiency win, not a handicap.** `se` at 30k beats `plain` at 100k (0.103 vs 0.310) using 3.3× less compute. At 2000 steps the ordering is reversed — which is a budget artifact, not an architecture property. I initially drew the wrong conclusion from the 2000-step data before running these.
2. **30k is not convergence.** 0.479 → 0.181 → 0.103 is still descending steeply, so even these are a lower bound.

## What this means for the scaling study

The x-axis in this issue is N, but the dominant variable is the step budget, and the two interact: at 2000 steps every N looks identical because nothing learns. **The N-sweep has to be re-run at ≥30k steps**, or it measures the wrong thing. The 120-cell run I did is only good as evidence that the original budget is uninformative.

That also re-prices everything. At 30k steps a 16-qubit quantum cell is ~42 h (measured), so 40 cells is ~1700 h. Options:

| config | quantum arm, 40 cells @ 30k |
|---|---|
| 8 qubits, `lightning.qubit` | **86 h** |
| 12 qubits, `lightning.qubit` | 275 h |
| 16 qubits, `lightning.gpu` | 1 682 h |

Dropping to 8 qubits keeps the ladder parameter-matched (`quantum − se = 48` = the circuit's own weights), so it stays a valid comparison — just a narrower bottleneck than the published 16, which would need stating. I'm currently checking whether `se` still learns at that width before recommending it.

## Incidental

`lightning.gpu` is 2–4× *slower* than CPU at 4–12 qubits (launch overhead) and 6× faster at 16. And it gives no benefit at all to sampling, which runs under `no_grad` — so it does not help the FID/KID cost, which is the larger half of the budget.
